@vates/nbd-client/index.js

```diff
@@ -78,24 +97,55 @@ module.exports = class NbdClient {
     })
   }
 
-  async connect() {
-    // first we connect to the serve without tls, and then we upgrade the connection
+  async #connect() {
+    // first we connect to the server without tls, and then we upgrade the connection
     // to tls during the handshake
     await this.#unsecureConnect()
     await this.#handshake()
-
+    this.#connected = true
     // reset internal state if we reconnected a nbd client
     this.#commandQueryBacklog = new Map()
     this.#waitingForResponse = false
   }
+  async connect() {
+    return pTimeout.call(this.#connect(), this.#connectTimeout)
+  }
 
   async disconnect() {
+    if (!this.#connected) {
+      return
+    }
+
     const buffer = Buffer.alloc(28)
     buffer.writeInt32BE(NBD_REQUEST_MAGIC, 0) // it is a nbd request
     buffer.writeInt16BE(0, 4) // no command flags for a disconnect
     buffer.writeInt16BE(NBD_CMD_DISC, 6) // we want to disconnect from nbd server
     await this.#write(buffer)
     await this.#serverSocket.destroy()
+    this.#serverSocket = undefined
+    this.#connected = false
+  }
+
+  #clearReconnectPromise = () => {
+    this.#reconnectingPromise = undefined
+  }
+
+  async #reconnect() {
+    await this.disconnect().catch(() => {})
+    await pDelay(this.#waitBeforeReconnect) // need to let the xapi clean things on its side
+    await this.connect()
+  }
+
+  async reconnect() {
+    // we need to ensure reconnections do not occur in parallel
+    if (this.#reconnectingPromise === undefined) {
+      this.#reconnectingPromise = pRetry(() => this.#reconnect(), {
+        tries: this.#reconnectRetry,
```

Comment:
IMHO, the name `reconnectRetry` is ambiguous.

`pRetry` supports both `tries` and `retries` options which are differs in that `tries = retries + 1`.

Also, it may be nice to allow other options (like `delay`) to be part of the configuration as well, as an example here's what I did in `@xen-orchestra/xapi`: 
https://github.com/vatesfr/xen-orchestra/blob/15f69a19f5ae013c0d64f81679ddd3d3851f6553/%40xen-orchestra/xapi/index.js#L153-L157